rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,518 @@
+
+The storage must also store the events of type `kind=state` since they are used during the CA rotation phase. They are helpful if the pod restarts, and has to rollback to the previous identity or finish the process and replace the old identity with the new one.
+
+### CA Rotation
+
+CA rotation feature allows the cluster operator to force the cluster to recreate and re-issue certificates for users and agents. During this procedure, agents and users receive the new keys signed by the newest CA from Teleport Auth.
+
+While the cluster is transitioning, the keys signed by the old CA are considered valid in order for the agent to be able to receive its new identity certificates from Teleport Auth. During the whole process, the CA rotation can be rollbacked, and if this happens the new certificate becomes invalid. This action can happen because another agent has issues or by decision of the operator. Given this, the new identity keys and old identity keys must be stored separately in the Secret until the rotation process finishes successfully.
+
+Once the new identity is received by the agent, it will store its state, indicating it is under a rotation event, and the new identity in the secret. The state under `/states/{role}/state` and the identity under `/ids/{role}/replacement`. This is mandatory so the pod, after restart, can resume the operation if not finished yet. If the process fails, the state is rewritten in order to inform the rotation has finished. Otherwise, if the process finishes successfully the process will replace the identity under `/states/{role}/current` with the new identity and replace the state content.
+
+### Upgrade from PRE-RFD versions
+
+When upgrading from PRE-RFD into versions that implement this RFD, we have three scenarios based on pre-existing configuration. They are described in the table below.
+
+| Storage | Replicas >1 | Scenario |
+|---|---|---|
+| true | false |  (1) |
+| true | true |  (1)  |
+| false | false | (2) |
+| false | true |  (3) |
+
+A description of each case's caveats is available below.
+
+1. Storage is available:  `Statefulset -> Statefulset`
+    
+    When storage is available, it means that identity and state were previously stored in the local database in PV. This means that the agent is able to read the local file and store that information in a new Kubernetes Secret, deleting the local database once the operation was successful.
+    
+    Once the secret is stored, the agent no longer requires the PV storage since nothing is stored there. At this point the operator can upgrade the Helm Chart to have `storage.enabled=false`.
+    
+    A descriptive comment and deprecation must be added to the storage section detailing the upgrade process. First, it's required in order to read the agent identity from local storage, and later it can be disabled.
+
+
+2. Storage is not available; Replicas = 1:  `Deployment -> Deployment`
```

Comment:
TBH I think we should treat `2` and `3` as the same scenario because the user can also use a `kubectl scale` command out-of-band to increase or decrease the number of replicas on a deployed cluster. So really we have 2 scenarios:

1. Persistent storage enabled. We're already using StatefulSet so we're good.
2. Persistent storage is not enabled. We're using Deployment which we need to migrate to StatefulSet.

For "2" - how does `helm upgrade` behave? If it first creates StatefulSet and then removes Deployment, then we should be good also. We'll just need to update our Helm chart to always create StatefulSet regardless of the storage setting.

If it first removes Deployment, and then creates StatefulSet, it will create a bit of downtime while the agents are recreated but might not be the end of the world also. We can release this in a minor or major release and mention the upgrade behavior in the release notes or something.

I would first experiment and see what the behavior is.

Reply:
I agree with merging `2` and `3`  into a single case, this means that we convert `Deployment -> Statefulset`.

Regarding the ordering, it first deletes the Deployment and after creates the Statefulset (it's the same order they use for [resource creation](https://github.com/helm/helm/blob/eeac83883cb4014fe60267ec6373570374ce770b/pkg/releaseutil/kind_sorter.go#L31)), but even if the opposite was happening, the downtime is almost certain since the pods take a while to be fully operational and there is no guarantee that the pods from statefulset are running before the deployment deletion takes place.